docs-issue-generation: Create docs issues in Jira

nickvigilante opened the pull request:
Fixes DOC-7066

Additional fixes:

- Update the authorization scheme for GitHub to `Bearer`
- Divide `docs_issue_generation.go` into logical package files
- Strip invalid epic references

Release note: None

rickystewart commented on pkg/cmd/docs-issue-generation/jira.go:
```diff
@@ -0,0 +1,147 @@
+// start time. For reference, it's structured as map[crdb_pr_number]map[crdb_commit]docs_pr_number.
+func searchJiraDocsIssues(startTime time.Time) (map[int]map[string]string, error) {
+	var result = map[int]map[string]string{}
+	pageSize, startAt := 100, 0
```
Nit: Prefer these assignments on 2 lines rather than the same line.

rickystewart commented on pkg/cmd/docs-issue-generation/github.go:
```diff
@@ -0,0 +1,294 @@
+func searchCockroachPRs(startTime time.Time, endTime time.Time) ([]cockroachPR, error) {
+	prCommitsToExclude, err := searchJiraDocsIssues(startTime)
+	if err != nil {
+		fmt.Println(err)
```
Is ignoring the problem and continuing for sure what we want to do? If so, I would appreciate a comment to clarify that.

nickvigilante replied:
That wasn't my original intention. I now return an error.

rickystewart commented on pkg/cmd/docs-issue-generation/github.go:
```diff
@@ -0,0 +1,294 @@
+	}
+	err := queryGraphQL(query, queryVariables, &search)
+	if err != nil {
+		fmt.Println(err)
```
Nit: since we're returning the error anyway, there's no reason to print it here. Someone else can handle it.

rickystewart commented on pkg/cmd/docs-issue-generation/github.go:
```diff
@@ -0,0 +1,294 @@
+	}
+	hasNextPage, nextCursor, prs, err := searchCockroachPRsSingle(startTime, endTime, "", prCommitsToExclude)
+	if err != nil {
+		fmt.Println(err)
```
As below: in `docs_issue_generation.go` line 68, the error is printed anyway. That means we don't have to repetitively print it here as well.

Same with the rest of the `fmt.Println(err)`: check whether they are actually needed. I would say that this is something you want to do *if and only if* we're just throwing away the error. If you're returning it, the job is done: someone else will handle it, print it, whatever.

nickvigilante replied:
I've updated it so we only print the error message within the `main` function in `main.go`. It returns it everywhere else.

rickystewart commented on pkg/cmd/docs-issue-generation/github.go:
```diff
@@ -0,0 +1,294 @@
+
+var searchCockroachReleaseBranches = func() ([]string, error) {
+	var result []string
+	hasNextPage, nextCursor, branches, err := searchCockroachReleaseBranchesSingle("")
```
It looks like we have a lot of small helper functions that are shaped like this: this one, `searchCockroachPRCommits`, `searchCockroachPRs`. It seems you can write the helper function *once*, and use generics to reduce the code duplication. Consider this optional since it's not too much duplicated code.

nickvigilante replied:
I think that'll be a good exercise for me as a follow-up. I'll need to think about this and strategize a bit.

rickystewart commented on pkg/cmd/docs-issue-generation/github.go:
```diff
@@ -0,0 +1,294 @@
+// getJiraIssueFromGitHubIssue takes a GitHub issue and returns the appropriate Jira key from the issue body.
+// getJiraIssueFromGitHubIssue is specified as a function closure to allow for testing
+// of getJiraIssueFromGitHubIssue* methods.
+var getJiraIssueFromGitHubIssue = func(org, repo string, issue int) (string, error) {
```
Nit: Bit weird to declare a top-level function like `var funcName = func(...)` Instead, I expect to see `func funcName(...)`.

nickvigilante replied:
The reason for this is because I need to "mock" network resources for the tests to run. Take a look at like 769 in `docs_issue_generation_test.go`: https://github.com/cockroachdb/cockroach/pull/109788/files#diff-7e172e15a18c5d651442463a3286130095be65025f92e406a75727201a198dc9R769

If you know of a way that I can mock the network resources inside of this function without needing to declare the function as a variable, I'm all ears.

rickystewart replied:
Hmm. For now, it's fine. The "proper" way to do this would be to define an `interface` and put this function behind that interface. Then you have one instance for "production", and one for testing that has your "mock" logic.

No need to change anything for now and increase the complexity of what you have.

nickvigilante replied:
Do you have an example in our codebase or elsewhere where that's done? I'll add that as a follow-up item for this ticket.

rickystewart replied:
Try an external resource like [this one](https://nathanleclaire.com/blog/2015/10/10/interfaces-and-composition-for-effective-unit-testing-in-golang/) or [this one](https://deliveroo.engineering/2019/05/17/testing-go-services-using-interfaces.html). It seems there are a ton of blog posts, if either of these don't make sense to you try Googling for them.

The short version is that this function would be "behind an interface" which might look like:

```
type jiraClient interface {
    getJiraIssueFromGitHubIssue(org, repo string, issue int) (string, error)
}

type productionJiraClient struct {}

func (_ productionJiraClient) getJiraIssueFromGitHubIssue(org, repo string, issue int) (string, error) {
   ....
}
```

The "production" `jiraClient` will just call `getJiraIssueFromGitHubIssue` as it's defined here. Any function that wants to call `getJiraIssueFromGitHubIssue()` needs a `jiraClient` instance. For testing, use a special instance that does whatever "mock" functionality you need.

rickystewart commented on pkg/cmd/docs-issue-generation/jira.go:
```diff
@@ -0,0 +1,147 @@
+		fmt.Println(err)
+		return err
+	}
+	//os.Stdout.Write(b.Bytes())
```
Delete line

rickystewart commented on pkg/cmd/docs-issue-generation/jira.go:
```diff
@@ -0,0 +1,147 @@
+	var b bytes.Buffer
+	encoder := json.NewEncoder(&b)
+	encoder.SetEscapeHTML(false)
+	err := encoder.Encode(dib)
```
Hmm. What's going on here? Why are we encoding the `docsIssueBatch` as JSON, then immediately de-coding it? What problem are we trying to solve here?

nickvigilante replied:
Yeah, that's pretty silly of me. I forgot that the `httpRequest` function already decodes the body with `SetEscapeHTML` to be `false`. I'll remove this blurb here.

rickystewart commented on pkg/cmd/docs-issue-generation/requests.go:
```diff
@@ -0,0 +1,127 @@
+}
+
+func httpRequest(
+	url, method, source string, headers map[string]string, body interface{}, out interface{},
```
Using a string for the `source` is not ideal as there are only two possible options: `GitHub` and `Jira`. What if I pass in something different? What if someone misspells them or forgets the correct capitalization?

Unfortunately Go doesn't support `enum`s because it's not a real programming language. However, you can kind of implement the same thing using `iota`.

```
type httpReqSource int
const (
    httpSourceGitHub httpReqSource = iota
    httpSourceJira httpReqSource
)
```

Then make `source` an `httpReqSource` instead of a `string`. To be safe, add a check to make sure `source` is either `httpSourceJira` or `httpSourceGitHub` and return an appropriate error if not.

rickystewart commented on pkg/cmd/docs-issue-generation/requests.go:
```diff
@@ -0,0 +1,127 @@
+	var tmp interface{}
+	err = json.Unmarshal(bs, &tmp)
+	if err != nil {
+		fmt.Printf("error: Unable to unmarshal JSON into an empty interface: %s", err)
```
`fmt.Errorf()` seems more applicable here, as in:

```
return fmt.Errorf("unable to unmarshal JSON (%w), byte slice: %s", err, string(bs))
```

nickvigilante commented on pkg/cmd/docs-issue-generation/extract.go:
```diff
@@ -0,0 +1,183 @@
+			Type: "text",
+			Text: "Epic:",
+		})
+		for x := range refInfo.epicRefs {
```
I agree. I first tried using `reflect` to do this, but it gave me issues when running tests. The function approach is a much better idea. I'll get started on this now. Thanks!

nickvigilante commented on pkg/cmd/docs-issue-generation/extract.go:
```diff
@@ -0,0 +1,183 @@
+	"strconv"
+)
+
+func parseDocsIssueBody(body string) (int, string, error) {
```
Eek, that function is only used in a test, which I need to update. I'm omitting that function in preference of the function below it.

nickvigilante commented on pkg/cmd/docs-issue-generation/docs_issue_generation.go:
```diff
@@ -204,520 +54,66 @@ var (
 	bugFixRNRE             = regexp.MustCompile(`(?i)release note:? \(bug fix\):.*`)
 	releaseJustificationRE = regexp.MustCompile(`(?i)release justification:.*`)
 	prNumberRE             = regexp.MustCompile(`Related PR: \[?https://github.com/cockroachdb/cockroach/pull/(\d+)\D`)
```
All variables moved closer to or in the same file where they're used.

Commit:
docs-issue-generation: Create docs issues in Jira

Fixes DOC-7066

Additional fixes:

- Update the authorization scheme for GitHub to `Bearer`
- Divide `docs_issue_generation.go` into logical package files
- Strip invalid epic references

Release note: None

rickystewart commented on pkg/cmd/docs-issue-generation/docs_issue_generation_test.go:
```diff
@@ -685,21 +3055,44 @@ Release note (bug fix): Fixin the bug`,
 
 	for _, tc := range testCases {
+		defer testutils.TestingHook(&getValidEpicRef, func(issueKey string) (bool, string, error) {
+			var epicMap = make(map[string]struct {
```
You can actually declare this in-line. It'll be way easier to read and write.

```
var epicMap = map[string]struct{
    IsEpic bool
    EpicKey string
} {
    "CRDB-491": {
        IsEpic: true,
        EpicKey: "CRDB-491"
     },
    "CRDB-9234": {
        IsEpic: false, EpicKey: "",
    }
    ...
```

rickystewart commented on pkg/cmd/docs-issue-generation/extract.go:
```diff
@@ -0,0 +1,234 @@
+}
+
+func containsEpicNone(message string) bool {
+	if allMatches := epicNoneRE.FindAllString(message, -1); len(allMatches) > 0 {
```
Hmm, I think this can just be
```
func containsEpicNone(message string) bool {
    return epicNoneRE.MatchString(message)
}
```

rickystewart commented on pkg/cmd/docs-issue-generation/format.go:
```diff
@@ -0,0 +1,277 @@
+	branchVersionRe := regexp.MustCompile(`release-(\d+\.\d+)`)
+	labelToLabelId := make(map[string]string)
+	for _, x := range fixVersions {
+		allMatches := labelNameRe.FindAllString(x.Name, -1)
```
As above, I think you can use `MatchString()` instead of `FindAllString()`.

rickystewart commented on pkg/cmd/docs-issue-generation/requests.go:
```diff
@@ -0,0 +1,131 @@
+		req.SetBasicAuth(jiraDocsUserEmail, tokenParams.JiraToken)
+	} else if source == httpSourceGitHub {
+		req.Header.Set("Authorization", githubAuthHeader)
+	}
```
Please add an `else { return fmt.Errorf("unexpected httpReqSource %d", source) }` or similar to protect if someone passes in some unexpected number.

nickvigilante commented:
bors r+
